### PR TITLE
fix!: retain checkbox group value on item label generator change

### LINKED ISSUE #3131: CheckboxGroup :  calling setItemLabelGenerator(...) in a ValueChangeListener unchecks all checkboxes
### Description

In a ValueChangeListener of a CheckboxGroup the label of the items are refreshed by calling setItemLabelGenerator(...).
When clicking on one of the checkboxes, the labels are changed, but the checkboxes are all unchecked afterwards.

### Expected outcome

I would expect, that the checked state of the checkboxes remain unchanged, exclusing the one clicked.

### Minimal reproducible example

[CheckboxGroupView.java2.txt](https://github.com/vaadin/flow-components/files/8632285/CheckboxGroupView.java2.txt)

### Steps to reproduce

Click one of the unchecked checkboxes

### Environment

Vaadin version(s):  23.0.6
OS: Windows 10

### Browsers

Chrome, Firefox

### PR DESCRIPTION
**Pull request by ugur-vaadin**
## Description

Calling `setItemLabelGenerator` for `CheckboxGroup` used to clear the selected values.

This PR makes the `CheckboxGroup` retain the selection by refreshing the checkboxes instead of resetting the `CheckboxGroup` when `setItemLabelGenerator` is called.

This is a breaking change.

Fixes #3131 

## Type of change

- [x] Bugfix
- [ ] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/contributing/overview
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

### CODE REVIEW

File: vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java [hunk 1]
```diff
@@ -484,7 +484,7 @@ public void setItemLabelGenerator(
         Objects.requireNonNull(itemLabelGenerator,
                 "The item label generator can not be null");
         this.itemLabelGenerator = itemLabelGenerator;
-        reset();
+        refreshCheckboxes();
```
Comment (tomivirkki) [inline vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java:487]: All the tests pass even without this addition. Could add a Junit test which also checks that the checkbox labels are updated.

Comment (tomivirkki) [inline vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java:487]: Should we align the implementation with [that of RadioButtonGroup](https://github.com/vaadin/flow-components/blob/b5a97698b1fc895bdb64d06c3e6b41abd8c38d0c/vaadin-radio-button-flow-parent/vaadin-radio-button-flow/src/main/java/com/vaadin/flow/component/radiobutton/RadioButtonGroup.java#L366)?

Comment (sissbruecker) [reply vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java:487]: Good point. The implementation in RBG seems to make more sense, but we should have a discussion about this in the team.

Ideally the JavaDocs should be improved as well, to mention that setting an item label generator overrides / removes the renderer.

Comment (ugur-vaadin) [reply vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java:487]: I tried again locally, seems to break `CheckboxGroupTest.selectItem_setItemLabelGenerator_selectionIsRetained` for me. I wonder why that could happen.

Comment (ugur-vaadin) [reply vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java:487]: I also added an integration test that checks if the labels are updated.

Comment (tomivirkki) [reply vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java:487]: > I tried again locally, seems to break

I meant the added line `+ refreshCheckboxes();`, not the entire change. The tests passed with `reset()` removed only.

Comment (ugur-vaadin) [reply vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java:487]: Oh, I get it. Then I suppose the new test addresses this concern.

Comment (tomivirkki) [reply vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java:487]: nit: an integration test might be a bit overkill for the purpose. Tried the following and it seems to cover it:

```java
Checkbox cb = (Checkbox) checkboxGroup.getChildren().findFirst().get();
Assert.assertEquals("foo (Updated)", cb.getLabel());
```

Comment (ugur-vaadin) [reply vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java:487]: I updated the PR to address this

Comment (ugur-vaadin) [reply vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java:487]: I will create a separate issue to align the behaviour of the components.

File: vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/test/java/com/vaadin/flow/component/checkbox/tests/CheckboxGroupTest.java [hunk 2]
```diff
@@ -302,6 +302,32 @@ public void allDataRefreshEvent() {
...
+        checkboxGroup.setValue(Set.of("foo"));
+        Assert.assertEquals(Set.of("foo"), checkboxGroup.getValue());
+
+        checkboxGroup.setItemLabelGenerator(item -> item + " (Updated)");
+
+        Assert.assertEquals(Set.of("foo"), checkboxGroup.getValue());
+    }
+
+    @Test
+    public void selectItem_setItemLabelGenerator_labelIsUpdated() {
```
Comment (sissbruecker) [inline vaadin-checkbox-flow-parent/vaadin-checkbox-flow/src/test/java/com/vaadin/flow/component/checkbox/tests/CheckboxGroupTest.java:319]:
```suggestion
    public void setItemLabelGenerator_labelIsUpdated() {
```
There isn't any selection happening in this test case.

**Review by tomivirkki — APPROVED** [review]

**Review by sissbruecker — APPROVED** [review]

**Pull request merged**